framework/src/modules/chain/dpos/apply.js

```diff
@@ -0,0 +1,211 @@
+
+		// When we need to sum round just after genesis block (height: 1)
+		// - set data manually to 0, they will be distributed when actual round 1 is summed
+		if (block.height === 1) {
+			return {
+				round,
+				totalFee: 0,
+				rewards: [0],
+				delegatePks: [block.generatorPublicKey],
+			};
+		}
+
+		try {
+			const row = await this.storage.entities.Round.summedRound(
+				round,
+				this.activeDelegates,
+				tx
+			)[0];
+
+			return {
+				round,
+				totalFee: new BigNum(Math.floor(row.fees).toPrecision(15)),
+				rewards: row.rewards.map(
+					reward => new BigNum(Math.floor(reward).toPrecision(15))
+				),
+				delegatePks: row.delegates,
+			};
+		} catch (err) {
+			this.logger.error('Failed to sum round', round);
+			this.logger.error(err);
+			throw err;
+		}
+	}
+
+	async getNonForgedDelegateAddresses(round, forgedDelegatesPks) {
+		const roundDelegatesPks = await this.delegates.generateActiveDelegateList(
+			round
+		);
+
+		return roundDelegatesPks
+			.filter(roundDelegatePk => !forgedDelegatesPks.includes(roundDelegatePk))
+			.map(delegate => getAddressFromPublicKey(delegate));
+	}
+
+	/**
+	 *  @todo `round` parameter is only necessary for handling
+	 * an exception in testnet in
+	 * `calculateRewardAndFeePerDelegate` method. `round` argument
+	 * can be safely removed when the exception on testnet was fixed.
+	 */
+	calculateRewardAndFeePerDelegate(
+		totalFee,
+		delegateReward,
+		isGettingRemainingFees,
+		round
+	) {
+		const exceptionRound = this.exceptions.rounds[round.toString()];
+		if (exceptionRound) {
+			// Apply rewards factor
+			delegateReward = delegateReward
+				.times(exceptionRound.rewards_factor)
+				.floor();
+
+			// Apply fees factor and bonus
+			totalFee = totalFee
+				.times(exceptionRound.fees_factor)
+				.plus(exceptionRound.fees_bonus)
+				.floor();
+		}
+
+		let fee = totalFee.dividedBy(this.activeDelegates).floor();
+		const reward = delegateReward.floor() || 0;
+
+		if (isGettingRemainingFees) {
+			const feesRemaining = totalFee.minus(
+				fee.times(this.scope.constants.activeDelegates)
+			);
+			fee = fee.plus(feesRemaining);
+		}
+
+		return {
+			fee,
+			reward,
+			amount: fee.plus(reward),
+		};
+	}
+
+	async distributeRewardsAndFees(
+		{ round, totalFee, rewards, delegatePks },
+		tx
+	) {
+		const updatedDelegates = delegatePks.map(
+			async (delegatePublicKey, index) => {
+				const address = getAddressFromPublicKey(delegatePublicKey);
+				const isGettingRemainingFees = index === delegatePks.length - 1;
+
+				const earnings = this.calculateRewardAndFeePerDelegate(
+					totalFee,
+					rewards[index],
+					isGettingRemainingFees,
+					round
+				);
+
+				const account = await this.storage.entities.Account.get(
+					{ address },
+					{ extended: true },
+					tx
+				);
+
+				const data = distributeRewardsAndFees(account, earnings);
+
+				await this.storage.entities.Account.update({ address }, data, tx);
```

Comment:
can we gather all the information first and do the update all at once in one place?